src/core/tsi/s2a/record_protocol/s2a_crypter.cc

```diff
@@ -598,6 +600,222 @@ grpc_status_code s2a_encrypt(s2a_crypter* crypter, uint8_t* plaintext,
   iovec plaintext_vec = {(void*)plaintext, plaintext_size};
   iovec record_vec = {(void*)record, record_allocated_size};
   return s2a_write_tls13_record(crypter, SSL3_RT_APPLICATION_DATA,
-                                &plaintext_vec, 1, record_vec, record_size,
-                                error_details);
+                                &plaintext_vec, /** plaintext_vec length **/ 1,
+                                record_vec, record_size, error_details);
+}
+
+grpc_status_code s2a_max_plaintext_size(const s2a_crypter* crypter,
+                                        size_t record_size,
+                                        size_t* plaintext_size,
+                                        char** error_details) {
+  GPR_ASSERT(crypter != nullptr);
+  GPR_ASSERT(plaintext_size != nullptr);
+
+  /** Note that we require this method returns 1 more than the size of the
+   *  plaintext that is returned by decrypting a TLS 1.3 record of size
+   *  |record_size|; this is because |crypter| requires 1 byte of extra space
+   *  after the plaintext to decrypt the record type. **/
+  size_t tag_size;
+  grpc_status_code tag_status = s2a_tag_size(crypter, &tag_size, error_details);
+  if (tag_status != GRPC_STATUS_OK) {
+    return tag_status;
+  }
+  GPR_ASSERT(record_size >= SSL3_RT_HEADER_LENGTH + tag_size);
+  *plaintext_size = record_size - SSL3_RT_HEADER_LENGTH - tag_size;
+  return GRPC_STATUS_OK;
+}
+
+grpc_status_code s2a_max_plaintext_size(const s2a_crypter* crypter,
```

Comment:
The arguments to the two `s2a_max_plaintext_size` methods have slightly different assumptions, because they have different callers. For this one, `|protected_vec|` is assumed to contain only the TLS payload, hence we should not subtract the header size. In the other method, we assume that `record_size` is the size of the entire TLS record.

Reply:
SG, thanks for the explanation. Please make sure this info is reflected in the function comments.

Reply:
Added another sentence to the comment about the function declaration to emphasize this.